File: java/src/main/java/com/benchmark/graph/Graph.java
package com.benchmark.graph;

import com.benchmark.models.Config;
import com.benchmark.models.Edge;
import com.benchmark.models.Node;
import com.benchmark.models.Vehicle;
import com.benchmark.physics.Physics;

import java.util.*;

/**
 * Weighted directed road network graph with pre-computed physical edge costs.
 *
 * <p>Constructs an adjacency-list representation over spatial vertices and directed
 * road segments. All edge physical metrics (aerodynamic drag, slope resistance, rolling
 * friction, fuel consumption, and per-mode costs) are evaluated once at construction
 * time via {@link Physics#computeEdgeMetrics}, rendering the graph read-only and safe
 * for concurrent access from multiple threads.
 *
 * <p>Adjacency entries are stored as {@link HalfEdge} instances containing a target
 * node ID and an index into the edge list, avoiding edge data duplication.
 */
public class Graph {

    /**
     * Compact outgoing edge entry in the adjacency list.
     *
     * <p>Stores a target vertex identifier and an index into the graph's edge list
     * to avoid memory duplication during adjacency construction.
     */
    public static class HalfEdge {
        /** Identifier of the destination vertex. */
        public int targetNode;
        /** Index of this edge in the parent graph's {@code edges} list. */
        public int edgeIndex;

        /**
         * Constructs a HalfEdge with the specified target and edge index.
         *
         * @param targetNode Destination vertex identifier.
         * @param edgeIndex  Index of the corresponding full Edge in the graph's edge list.
         */
        public HalfEdge(int targetNode, int edgeIndex) {
            this.targetNode = targetNode;
            this.edgeIndex = edgeIndex;
        }
    }

    /** Maps vertex identifier to its spatial Node geometry. */
    public Map<Integer, Node> nodes;
    /** Ordered list of all directed road segments with synthesized cost fields. */
    public List<Edge> edges;
    /** Vehicular physical constants for this benchmark run. */
    public Vehicle vehicle;
    /** Atmospheric and optimization weight configuration parameters. */
    public Config config;
    /** Forward adjacency index: {@code node_id -> List<HalfEdge>}. */
    public Map<Integer, List<HalfEdge>> adj;

    /**
     * Constructs the graph, executes full physics derivation on all edges,
     * and builds the forward adjacency index.
     *
     * @param nodesList Complete list of spatial graph vertices.
     * @param edgesList Complete list of directed road segments (mutated in-place).
     * @param vehicle   Vehicular aerodynamic and thermodynamic constants.
     * @param config    Atmospheric parameters and multi-objective weight coefficients.
     */
    public Graph(List<Node> nodesList, List<Edge> edgesList, Vehicle vehicle, Config config) {
        this.vehicle = vehicle;
        this.config  = config;
        this.edges   = edgesList;
        this.nodes   = new HashMap<>(nodesList.size());
        this.adj     = new HashMap<>(nodesList.size());

        for (Node n : nodesList) {
            this.nodes.put(n.id, n);
            this.adj.put(n.id, new ArrayList<>());
        }

        for (int i = 0; i < edgesList.size(); i++) {
            Edge e = edgesList.get(i);
            Physics.computeEdgeMetrics(e, vehicle, config);
            List<HalfEdge> neighbors = this.adj.get(e.source);
            if (neighbors != null) {
                neighbors.add(new HalfEdge(e.target, i));
            }
        }
    }

    /**
     * Returns the pre-computed objective cost for the specified optimization mode.
     *
     * @param edge Road segment whose cost is to be selected.
     * @param mode Optimization strategy: {@code "FASTEST"}, {@code "CHEAPEST"}, or {@code "BALANCED"}.
     * @return Scalar edge traversal cost for priority queue relaxation.
     */
    public double getEdgeWeight(Edge edge, String mode) {
        if ("FASTEST".equals(mode)) {
            return edge.costFastest;
        } else if ("CHEAPEST".equals(mode)) {
            return edge.costCheapest;
        } else {
            return edge.costBalanced;
        }
    }
}

File: go/pkg/graph/graph.go
// Package graph constructs and indexes the weighted directed road network.
//
// Builds a forward adjacency structure from raw vertex and edge datasets,
// executing full physics evaluation on each edge during construction.
// The resulting Graph is read-only after NewGraph returns, making it safe
// for concurrent access across multiple goroutines without additional locking.
package graph

import (
	"route_optimization_go/pkg/models"
	"route_optimization_go/pkg/physics"
)

// HalfEdge encodes a compact outgoing edge entry in the adjacency list.
// Stores the target node identifier and an index into the Graph's edge slice
// to avoid duplicating edge data in memory.
type HalfEdge struct {
	// TargetNode is the identifier of the destination vertex.
	TargetNode int
	// EdgeIndex is the position of this edge in Graph.Edges.
	EdgeIndex int
}

// Graph is a weighted directed road network with pre-computed physical edge costs.
//
// Fields are exported to allow read access from pathfinding algorithms without
// requiring accessor methods, keeping hot-path code allocation-free.
type Graph struct {
	// Nodes maps vertex identifier to spatial Node geometry.
	Nodes map[int]models.Node
	// Edges is the ordered slice of all directed road segments.
	Edges []models.Edge
	// Vehicle holds the vehicular physical constants for this benchmark run.
	Vehicle models.Vehicle
	// Config holds atmospheric parameters and optimization weight coefficients.
	Config models.Config
	// Adj is the forward adjacency index: node_id -> []HalfEdge.
	Adj map[int][]HalfEdge
}

// NewGraph constructs the Graph, runs physics derivation on all edges,
// and builds the forward adjacency index.
//
// Parameters:
//
//	nodes   - Complete list of spatial graph vertices.
//	edges   - Complete list of directed road segments (modified in-place).
//	vehicle - Vehicular aerodynamic and thermodynamic constants.
//	config  - Atmospheric parameters and multi-objective weight coefficients.
//
// Returns: Pointer to a fully constructed, read-only Graph.
func NewGraph(nodes []models.Node, edges []models.Edge, vehicle models.Vehicle, config models.Config) *Graph {
	nodeMap := make(map[int]models.Node, len(nodes))
	adj := make(map[int][]HalfEdge, len(nodes))

	for _, n := range nodes {
		nodeMap[n.ID] = n
		adj[n.ID] = make([]HalfEdge, 0, 4)
	}

	for i := range edges {
		physics.ComputeEdgeMetrics(&edges[i], &vehicle, &config)
		adj[edges[i].Source] = append(adj[edges[i].Source], HalfEdge{
			TargetNode: edges[i].Target,
			EdgeIndex:  i,
		})
	}

	return &Graph{
		Nodes:   nodeMap,
		Edges:   edges,
		Vehicle: vehicle,
		Config:  config,
		Adj:     adj,
	}
}

// GetEdgeWeight returns the pre-computed objective cost for the specified optimization mode.
//
// Parameters:
//
//	edge - The road segment whose cost is to be selected.
//	mode - Optimization strategy: "FASTEST", "CHEAPEST", or "BALANCED".
//
// Returns: Scalar edge traversal cost for priority queue relaxation.
func (g *Graph) GetEdgeWeight(edge *models.Edge, mode string) float64 {
	switch mode {
	case "FASTEST":
		return edge.CostFastest
	case "CHEAPEST":
		return edge.CostCheapest
	default:
		return edge.CostBalanced
	}
}
